packages/edit-site/src/components/add-new-template/utils.js

```diff
@@ -0,0 +1,132 @@
+		( { viewable, slug } ) =>
+			viewable && ! excludedPostTypes.includes( slug )
+	);
+	return filteredPostTypes;
+};
+
+/**
+ * @typedef {Object} PostTypeEntitiesInfo
+ * @property {boolean}  hasEntities   If a postType has available entities.
+ * @property {number[]} existingPosts An array of the existing entities ids.
+ */
+
+/**
+ * Helper hook that returns information about a post type having
+ * posts that we can create a specific template for.
+ *
+ * First we need to find the existing posts with an associated template,
+ * to query afterwards for any remaing post, by excluding them.
+ *
+ * @param {string[]} existingTemplates The existing templates.
+ * @return {Object<string,PostTypeEntitiesInfo>} An object with the postTypes as `keys` and PostTypeEntitiesInfo as values.
+ */
+export const usePostTypesEntitiesInfo = ( existingTemplates ) => {
+	const postTypes = usePostTypes();
+	const slugsToExcludePerEntity = useMemo( () => {
+		return postTypes?.reduce( ( accumulator, _postType ) => {
+			const slugsWithTemplates = ( existingTemplates || [] ).reduce(
+				( _accumulator, existingTemplate ) => {
+					const prefix = `single-${ _postType.slug }-`;
+					if ( existingTemplate.slug.startsWith( prefix ) ) {
+						_accumulator.push(
+							existingTemplate.slug.substring( prefix.length )
+						);
+					}
+					return _accumulator;
+				},
+				[]
+			);
+			if ( slugsWithTemplates.length ) {
+				accumulator[ _postType.slug ] = slugsWithTemplates;
+			}
+			return accumulator;
+		}, {} );
+		// It's important to use `length` as a dependency because `usePostTypes`
+		// returns a new array every time and will triger a rerender.
+		// We can't avoid that because `post types` endpoint doesn't allow filtering
+		// with `viewable` prop right now.
+	}, [ postTypes?.length, existingTemplates ] );
+	const postsToExcludePerEntity = useSelect(
+		( select ) => {
+			if ( ! slugsToExcludePerEntity ) {
+				return;
+			}
+			const postsToExclude = Object.entries(
+				slugsToExcludePerEntity
+			).reduce( ( accumulator, [ slug, slugsWithTemplates ] ) => {
+				const postsWithTemplates = select( coreStore ).getEntityRecords(
+					'postType',
+					slug,
+					{
+						_fields: 'id',
+						context: 'view',
+						slug: slugsWithTemplates,
+					}
+				);
+				if ( postsWithTemplates?.length ) {
+					accumulator[ slug ] = postsWithTemplates;
+				}
+				return accumulator;
+			}, {} );
+			return postsToExclude;
+		},
+		[ slugsToExcludePerEntity ]
+	);
+	const entitiesInfo = useSelect(
+		( select ) => {
+			return postTypes?.reduce( ( accumulator, { slug } ) => {
+				const existingPosts =
+					postsToExcludePerEntity?.[ slug ]?.map(
+						( { id } ) => id
+					) || [];
+				accumulator[ slug ] = {
+					hasEntities: !! select( coreStore ).getEntityRecords(
+						'postType',
+						slug,
+						{
+							per_page: 1,
+							_fields: 'id',
+							context: 'view',
+							exclude: existingPosts,
+						}
+					)?.length,
```

Comment:
I tried `hasEntityRecords` but it doesn't make the requests. I think it's something more internal that only search in cache? 🤔 Any insights @adamziel maybe?

Reply:
`hasEntityRecords` does almost the same thing as you do here:

https://github.com/WordPress/gutenberg/blob/33d84b036592a5bf31af05b7710f3b2b14163dc4/packages/core-data/src/selectors.ts#L470-L477

It won't make the request if, say, there's a typo in `kind` or `slug` but should be good otherwise. Would you be able to reproduce the problem in a test case?

Reply:
Actually, I don't think running the selector internally triggers the resolvers! Your code goes through the registry where the selectors are bound to resolvers so that's why it works.

Reply:
So, it sounds like `hasEntityRecords` only checks if the entity records were already resolved – it's even the documented behavior:

> Returns true if records have been received for the given set of parameters,
or false otherwise.

This is quite confusing! Given that, your implementation looks fine.

Reply:
Thanks for chiming in.

Reply:
Thanks for explaining Adam!